soroban-env-host/src/host/error.rs

```diff
@@ -109,6 +109,32 @@ impl HostError {
             }
         }
     }
+
+    /// Identifies whether the error can be meaningfully recovered from.
+    ///
+    /// We consider errors that occur due to broken execution preconditions (
+    /// such as incorrect footprint) non-recoverable.
+    pub fn is_recoverable(&self) -> bool {
+        // All internal errors that originate from the host can be considered
+        // non-recoverable (they should only appear if there is some bug in the
+        // host implementation or setup).
+        if !self.error.is_type(ScErrorType::Contract)
+            && !self.error.is_type(ScErrorType::WasmVm)
```

Comment:
Why should we allow recovering from WasmVm InternalErrors?

Reply:
I've audited all the other error types and I'm reasonably certain that internal errors represent bugs. However, for wasm we return an internal error for any wasmi error that we can't simply convert to an error code (which IIUC is majority of wasmi errors). I wouldn't want it to be an internal error, but I don't know what it should be. I guess alternative could be to add one more error code specifically for broad wasmi errors... FWIW the risk of encountering an internal error is hopefully low (as they shouldn't happen unless we have a bug), so maybe it's not a big deal. Also we have just 2 or 3 more places where we return wasmvm internal error; maybe these should be turned into context errors instead?

Reply:
Why don't you want it to be an internal error? We don't expect them to happen right? And if they do, we don't want users assuming that an entry is missing when instead something weird happened in wasmi.

Reply:
I'm a little worried the rule of which errors are recoverable vs non-recoverable is not clear enough, and I feel basing it on a "blacklist" approach on certain types or codes may create more footgun.
In many places we are remapping an error type into a different error type. So depending on if the wrapping function is called or if the wrapped code is called from another path, the error might be different even though it originates from the same logic. This just may redefine whether the logic is resumable or not by surprise. 
Also to mention the error category and codes are "best effort", not perfect and may evolve. 
There should be a more thorough mechanism for recoverability I feel, maybe redesign parts of try_call or host invocation, I don't have good suggestions right now.

Reply:
> Why don't you want it to be an internal error? We don't expect them to happen right? 

Currently some errors that just happen naturally during wasm execution (like overflow or something like that) would end up reported as internal errors. I agree that we could try given them a different name and leave internal errors to host errors only. Looking for suggestions as to what these should be though.

> In many places we are remapping an error type into a different error type.

I'd say it's a bug if we wrap an internal error somewhere. I haven't noticed anything like that, but I didn't look for such scenarios though...

> Also to mention the error category and codes are "best effort", not perfect and may evolve.

That's fine if these evolve. This PR focuses on very specific aspects: resource errors and internal errors which would be `releaseAssertOrThrow` in core. I agree that we could handle both of these differently, but that's just a quick way to achieve that. It shouldn't be too hard to maintain backwards compatibility with these, unless we think that it's probably that we're going to get a lot of internal errors (again, I would really hope these never happen). But if we're too concerned about that, I'd say let's just make only resource errors non-recoverable (these are really narrow and straightforward).

> There should be a more thorough mechanism for recoverability I feel, maybe redesign parts of try_call or host invocation, I don't have good suggestions right now.

Honestly, I feel like the whole `try_call` thing is a footgun and I would be happier if we had neither this function nor rollbacks. It basically only seems suitable for batching completely unrelated operations together. But I guess we have to support this now, so it would be nice to remove at least some non-obvious footguns.